### ISSUE TITLE
Very slow loading and loop after update to Firefox v100

### ISSUE THREAD

**Issue by saelic** [piroor/treestyletab#3130]
Hello,

after upgrading to Firefox v100.x the TST has a problem with loading 6000+ tabs. The list goes into a loop i.e. it tries to load the tabs, but after some time it starts from the beginning (which is noticeable by list blinking). The workaround for this is to disable the theme ("No decoration"), but still it can loop sometimes.

Anyway, the performance with some many tabs is terrible, even on Intel Core i7. I know that it is absurd to have so many tabs open, but I also think that it is absurd that nowadays so many things are done via browser, so the mess can emerge very quickly. Optimization of tabs loading in such scenarios is a must.

**Comment by irvinm** [comment]
You will probably want to read over https://github.com/piroor/treestyletab/issues/3012 to see what information you can provide around your specific situation to progress this issue.

**Comment by saelic** [comment]
It seems that the problem is connected to some changes in Firefox's sandboxing mechanism.
During TST loading, the Threads look like this:
![treestyletab](https://user-images.githubusercontent.com/13816565/172929735-d115d6c1-c66b-4af7-982d-fc67bdca7118.png)

It then jumps to execute the _xul.dll_ library, but then goes back to _IsSandboxedProcess_ thread.

The problem seems to be similar to this:
https://www.reddit.com/r/pockettube/comments/tnqbu1/pt_freezes_firefox_requiring_me_to_closerestart_ff/

Please take a look at it. It drives me crazy to wait 20-30 MINUTES to load the tree - and even sometimes it loops indefinitely.

**Comment by saelic** [comment]
@piroor Please take a look at it, this is really annoying bug which makes Firefox unusable.

**Comment by tstackhouse** [comment]
I'm seeing this lag a lot on Linux, it can take 10-20 seconds or more to switch to a new tab, but it's not something that happens consistently.  And I only usually have 10-30 tabs open, not the hundreds that some users are reporting.

**Comment by saelic** [comment]
Guys, FOR THE LOVE OF GOD, **PLEASE TAKE A LOOK AT THIS PROBLEM**. Can you imagine how disturbing it is to wait **every single day for 30-60 MINUTES to open f*cking web browser???** **We are going crazy here, this is research facility, we can't do our tasks!** We are using high-end workstations with Xeon Processors and gigabytes of RAM and **we can't load stupid web browser!**

I know that this problem emerged because Mozilla messed something up in version 100 (yes, our mistake that we didn't go with Firefox ESR), but PLEASE at least take a look at it! @piroor @irvinm 

Problem also reported to Mozilla here: https://bugzilla.mozilla.org/show_bug.cgi?id=1774949

**Comment by irvinm** [comment]
1. I am just a user like you ... just trying to help out.
2. You haven't included any simplified steps on how to reproduce ... have you tried a new profile with just TST and your session?
3. You haven't listed your configuration.  "i7" with "Gigabytes" of ram isn't useful.
4. I created a new profile, added TST and Tab Session Manager, restored my average 500 tab session many times to get up to around 7000 tabs and was able to start up my browser to a working state in 2 minutes 30 seconds.  My machine is from 2014 with an Intel i7-4790k with 32GB of ram. (OLD!)
5. My average session with around 500 tabs starts up in under 20 seconds.

7000 tabs (TST and TSM only) - 2 min 30 seconds:
https://user-images.githubusercontent.com/979729/174462283-d3ad5d1b-8547-4162-a9b0-fdd7fa015e25.mp4

500 tabs (TST + 64 other addons) - <20 seconds: 
https://user-images.githubusercontent.com/979729/174462296-79f25e7e-8d18-45dc-b54a-fb2b334def59.mp4

This all seems reasonable based on machine performance and the size of the session.

**Comment by piroor** [comment]
I usually use Firefox + TST with 500 or less tabs without such a trouble, and I have no plan to use 1000 or more tabs on my daily use. Present performance of TST is enough for me, thus I have less motivation to tune TST up for huge number of tabs. I know the importance of performance, but I can't spend time to improve TST too much beyond my personal use, due to my other various tasks to live.

Patches for performance improvements are always welcome. I hope someone contributes to this project to improve its performance.

**Comment by saelic** [comment]
Hi guys, thank you for your attention.

The problem is that we are not talking about general performance improvements, but about **a potential bug that emerged with Firefox version 100** which causes TST (and other add-ons, as shown via attached links) to take abnormal time period to load.
We did NOT have any performance problems with TST **until Firefox updated itself to version 100** - I clearly stated this in the topic of this issue.

I've attached above a screenshot from Process Explorer which shows, that Firefox loops itself on one of it's threads. It is probably related to some security mechanism - as we can say based on thread name. I don't know what I can do more to trace the origin of the problem.

**Comment by irvinm** [comment]
@saelic the best way to address this issue as you believe it is related to changes in Firefox is to use [mozregression](https://mozilla.github.io/mozregression/) to trace back to the check-in that caused the change.  This is the information you will need to include in the bug you reported on bugzilla.

It is a pretty cool utility that allows you to run your current profile on different versions of Firefox to determine what check-in changed the behavior.  If you believe it was FF100 related, you can focus on a narrow timeline to find the change ... which shouldn't take too long as the tool takes care of most of the heavy lifting for you.

**Comment by saelic** [comment]
New findings: **the number of open tabs is NOT the cause of this bug.**

1. If we disable the navigation panel (so the tree is not visible), then go to TST options -> Advanced and uncheck or check "Optimize tree restoration with cache", then we encounter the same loop-like behaviour with IsSandboxedProcess thread, as shown above https://github.com/piroor/treestyletab/issues/3130#issuecomment-1151547983
So when this loop goes on, we are **probably** encountering some race-condition which makes loading the tree of tabs a totally random possibility.

2. We've restored an old profile from some months ago and - with Firefox 101 and old version of TST - the tree loaded normally.

So the conclusion is, that **the problem lies around sandboxing mechanism in Firefox and TST**.
Using the check-uncheck method described in point 1. can help with debugging.

**Comment by SXZ1** [comment]
TL;DR - use TST 3.4.10 as it doesn't suffer from this bug with whatever Firefox version you like including the latest Nightly (22-10-22), I will likely find the regression commit later, the only thing I currently got is a bunch of errors from the browser console.

Full version:

I've just realized I never come across this bug in the latest Nightly (22-10-22) with my main profile with a ~9000 tabs open in a session that I regularly restore every couple of days. The key to success is using TST3.4.10, it is available on Mozilla's website:
https://addons.mozilla.org/en-US/firefox/addon/tree-style-tab/versions/

Keep in mind installing it on top of the later versions will brake some settings like theme settings but in my experience, it doesn't do anything that can't be fixed by going into TST options and configuring everything over again but still make a copy of your profile just in case. Css code will probably need tweaking too if you use any. 

As a bonus, this version also expands/collapses tab stacks by 5-10 times faster compared to the latest available - for example, 3.4.10 expands 2000 tab stack for under a second while 3.9.7 takes around 3-5 seconds to do the same thing, that's quite an improvement and that's the reason I've been using 3.4.10 all this time, actually.

Funny thing is, I tried to reproduce this loop bug shortly after it was reported but came to a conclusion I can't draw any conclusion because the results I was getting were very inconsistent. Now I understand why - most likely uMatrix I was using alongside TST somehow interfered with the TST update process so considering I am now aware of that and I have the last known "good" TST revision to work with, I am sure I will find the regression commit as soon as I find space on my drive to install Ubuntu or buy a new drive. The only thing I have for now is a bunch of errors from the browser console.

**Comment by SXZ1** [comment]
I should have clarified I got two profiles with TST3.4.10 and TST3.9.7 respectively and I can consistently and reliably reproduce this bug in the profile with TST3.9.7 installed yet I've never come across this bug in profile with TST3.4.10.
@saelic Could you try 3.4.10 and give us feedback?

**Comment by saelic** [comment]
@SXZ1 Hello, thank you for your work, I'm glad that you've encountered the same problem. For this moment I am unable to check this.

Keep in mind that this problem could be somehow connected to the Firefox security sandboxing mechanism - as it is shown on the screenshot here: https://github.com/piroor/treestyletab/issues/3130#issuecomment-1151547983

I have no idea how, but I am almost certain that we are encountering some deadlock / internal loop while checking "something", as it is clearly shown on thread list. The xul.dll threads doesn't do much after enabling TST addon, as they are probably waiting for firefox.exe!IsSandboxedProcess thread to finish. I believe that this is a place to start when looking up to fix this bug.

The interesting thing is that this check (firefox.exe!IsSandboxedProcess) is simultaneously done by two separate Firefox processes - by the parent process and one child process.

Keep in mind, that the same problem encounters when ticking/unticking the "Optimize tree restoration with cache". After ticking the same "firefox.exe!IsSandboxedProcess" loop takes place, so maybe this is somehow connected to disk I/O - I'm just speculating, but after unticking and reticking this checkbox the tree restoration takes reasonable time until next session.

**Comment by irvinm** [comment]
@saelic I do not believe this is related to Firefox sandboxing.  My performance is good and the numbers I quoted in my previous reply shows similar percentage of calls to IsSandboxedProcess but the overall time is good.

Here is what I just did by hiding the sidebar and showing again ...

![image](https://user-images.githubusercontent.com/979729/197395775-e4f91cdd-ccba-448f-bc3b-f1b903227a98.png)

I also see that you reported to defect to Mozilla and they were not able to reproduce it and they closed it.

You could also try (as a debug step) relaxing the sandboxing levels to see how that impacts performance.  I would not recommend doing that long term.  (https://wiki.mozilla.org/Security/Sandbox)

**Comment by piroor** [comment]
If the change of sandbox is identified which cause this problem, it possibly be a help to do something workaround. I put this open again for further investigation by contributors.

**Comment by saelic** [comment]
@irvinm So look at this: two different processes hang on the same thread while loading TST.
![treestyletab 2](https://user-images.githubusercontent.com/13816565/197897108-9a4c82bb-f905-4ed8-a905-cd17b433acf7.png)

**Comment by SXZ1** [comment]
> considering I am now aware of that and I have the last known "good" TST revision to work with, I am sure I will find the regression commit as soon as I find space on my drive to install Ubuntu or buy a new drive

And here I am half a year later with a regression commit. Sorry it took so long, guys, my depression makes it very difficult to do even simple things.

Last good commit when bug is NOT reproduced: 710f3a551ed0c33066ab9dc57f03774626df7c49
First bad commit when bug IS reproduced: 748477a29b8b5babea7b7b53cb23d946ae799857
I've just manually set  **connectionTimeoutDelay** back to 5000 in my regular profile with 6700 tabs and the problem went away: 3 failed session restore attempts before the fix, 3 successful session restore attempts after the  fix.

To find the regression commit, I used a new profile with the only change being "Restore previous session automatically at startup" set to enabled or whatever that option is called.

@irvinm 

> 7000 tabs (TST and TSM only) - 2 min 30 seconds:

And this comment makes me wonder why this bug only occurs on certain computers. Interesting.

**Comment by SXZ1** [comment]
I don't remember if anyone hears my scream in a stale issue so tagging @piroor just in case.

**Comment by saelic** [comment]
I hear you, thank you very much for your work. Also sorry to hear about your mental state, keep in mind that you are not alone, life is getting harder and harder to live. Gosh I wish to go back to the 90's, or 2004 max.

> And this comment makes me wonder why this bug only occurs on certain computers. Interesting.

From the beginning I think, that this is somehow connected to the 2018's new CPU attacks and related new Intel Microcode versions and Windows Updates addressing them. That's only my speculaction/feeling.

**Comment by eshate** [comment]
Hey, I found this thread while looking for an answer to why it's taking so long to load. Something is definitely on the table. I'll be happy to provide some logs, but I need your help, sorry. How can I effectively log this issue to make it reproducible - and what information would be considered useful?

What I can tell for now is that:

1. I have two instances of 111.0 FFDev - one on PopOS and one on Windows, as I'm dual-booting.
2. A few months ago, an update destroyed my userChrome settings, and I found time to take care of it just a few days ago. I started with Windows, and here I'm facing these issues.
3. Pop is still working without TST and works fine.
4. I'm also using SimpleTabGroups, but all of my tabs are frozen there, so there's nothing running in the background. While opening FF, there are about 20 tabs on the TST panel, and even if only two tabs are actively loading, I still need to wait about one minute to start browsing.

For now, I don't know what else to write. I'm a little dizzy (and busy) now. Please let me know how I can contribute and with what info. Appreciate it.

**Comment by saelic** [comment]
@piroor Can we expect that this fix will be applied to the next version of TST?

**Comment by SXZ1** [comment]
Since we've summoned **Piroor**, here's the summary of the thread:

Short description of the problem from the 1st post:
>TST has a problem with loading 6000+ tabs. The list goes into a loop i.e. it tries to load the tabs, but after some time it starts from the beginning (which is noticeable by list blinking)


Comment that proves this bug doesn't manifest itself at least on some computers even with the extreme number of tabs:
https://github.com/piroor/treestyletab/issues/3130#issuecomment-1159597786
>7000 tabs (TST and TSM only) - 2 min 30 seconds:

Finally, regression commit:
https://github.com/piroor/treestyletab/issues/3130#issuecomment-1453973500
>Last good commit when bug is NOT reproduced: https://github.com/piroor/treestyletab/commit/710f3a551ed0c33066ab9dc57f03774626df7c49
First bad commit when bug IS reproduced: https://github.com/piroor/treestyletab/commit/748477a29b8b5babea7b7b53cb23d946ae799857
I've just manually set connectionTimeoutDelay back to 5000 in my regular profile with 6700 tabs and the problem went away

**Comment by piroor** [comment]
As a workaround I've introduced a change bb36b27 to increase the timeout seconds automatically. I hope it make things better...

**Comment by saelic** [comment]
Thank you! Can't wait to test it out on our cases.

**Comment by fmjrey** [comment]
> As a workaround I've introduced a change [bb36b27](https://github.com/piroor/treestyletab/commit/bb36b27c0eb427548e6282f5ea37640732d4d934) to increase the timeout seconds automatically. I hope it make things better...

I may not have 6000+ tabs, maybe a 1000, nevertheless after loading the dev build just now my life as a FF+TST user has become so much lighter, finally!

Thanks to all who contributed.

**Comment by saelic** [comment]
Glad to hear that!

@piroor Could you please release a 3.9.15 version which has this fix applied? Don't want to mess around with dev builds, this issue is so important, in progress since May, and as we hear - the workaroud worked.

Btw: why you are calling it a workaround? Isn't your commit a final fix for this problem?

**Comment by fmjrey** [comment]
Hmm,this may or may not be related to the fix: I wanted to move a tree of about 15 tabs or so from one window to another and somehow it broke TST. After waiting for a while maybe a minute or so, displaying history then reverting to TST ended up showing a blank space where the tree is normally displayed. I quit FF but when restoring the previous session the target window was lost, it wasn't listed in the history, so its content was lost :(
I can't reproduce this issue though, trying again to move the tabs to a newly created window worked in a flash. So that's why I can't say for sure it's related to the fix.
Edit: It happened again, small tree, dragging the head to another window made TST unable to update its display. This time I killed FF instead of quitting and all windows seem to be recovered upon restart. I was able to move around a few trees before it happened again.

**Comment by Milor123** [comment]
> 1. I am just a user like you ... just trying to help out.
> 
>     2. You haven't included any simplified steps on how to reproduce ... have you tried a new profile with just TST and your session?
> 
>     3. You haven't listed your configuration.  "i7" with "Gigabytes" of ram isn't useful.
> 
>     4. I created a new profile, added TST and Tab Session Manager, restored my average 500 tab session many times to get up to around 7000 tabs and was able to start up my browser to a working state in 2 minutes 30 seconds.  My machine is from 2014 with an Intel i7-4790k with 32GB of ram. (OLD!)
> 
>     5. My average session with around 500 tabs starts up in under 20 seconds.
> 
> 
> 7000 tabs (TST and TSM only) - 2 min 30 seconds: https://user-images.githubusercontent.com/979729/174462283-d3ad5d1b-8547-4162-a9b0-fdd7fa015e25.mp4
> 
> 500 tabs (TST + 64 other addons) - <20 seconds: https://user-images.githubusercontent.com/979729/174462296-79f25e7e-8d18-45dc-b54a-fb2b334def59.mp4
> 
> This all seems reasonable based on machine performance and the size of the session.

@irvinm What is the solution for slow charging tabs in out i7 4780k processor? I have a rx6750 xt. And firefox goes very slow try charging the tab for first time.

**Comment by saelic** [comment]
Hello @Milor123, we are now waiting for @piroor to release new version of TST which will have the applicable patch/workaround for this problem and will see how it goes.

**Comment by SXZ1** [comment]
@saelic update with a fix is already available on AMO.

**Comment by saelic** [comment]
I'm sorry, but what is AMO? :)

**Comment by SXZ1** [comment]
@saelic 
addons.mozilla.org, official Mozilla addon repository. In other words, TST should soon update to the fixed version on its own but you can speed up the process by clicking on the hamburger icon (three horizontal lines at the right end of the navigational bar) - Addons and themes - click on the gear - Check for updates.

TST will then update to the version with a bug fix. Better do that when TST sidebar is not loading.

**Comment by saelic** [comment]
@SXZ1 Thank you! I've checked for updates from Add-ons config page in Firefox and it found an update, thank you for pointing this out for me :)

**And great news! After update my ~8000 tabs loaded in about 1-2 minutes on the same computer! And it finally doesn't loop all the time while loading!**

**I'M SO GLAD! HUGE THANK YOU TO EVERYONE INVOLVED IN THIS CASE! This fix really has direct impact on my day-to-day ussage of Firefox and I'm sure there are others like me.**

Wish you all the best. I think that this bug is finally resolved after almost a year of investigation.
